### PR TITLE
fix(models): add Codex manifest fallback

### PR DESCRIPTION
**Pull request by Brisbanehuang**
## Summary

- return a Codex-compatible `{"models":[]}` envelope for authenticated `GET /v1/models` requests with a non-empty `client_version`
- add a stable weak ETag, standard `If-None-Match` handling, and `Cache-Control: no-cache`
- preserve the existing OpenAI `{"object":"list","data":[...]}` response for all other model-list requests

## Why

Codex manifest consumers use `client_version` to distinguish their model-catalog request and require a top-level `models` array. CCH currently routes these requests through its OpenAI model aggregator, so strict gateways reject the `{object,data}` response as an invalid manifest.

The empty manifest is intentional. Codex falls back to its bundled model catalog when the remote manifest is empty, retaining the client's complete capability metadata. CCH's aggregated model shape only contains a small subset of that metadata, so synthesizing non-empty entries would be unsafe. See the upstream Codex regression coverage: https://github.com/openai/codex/blob/44918ea10c0f99151c6710411b4322c2f5c96bea/codex-rs/models-manager/src/manager_tests.rs#L592-L608

Authentication remains ahead of the compatibility response, so missing, invalid, disabled, or expired credentials keep their existing `401` behavior.

**Related Issues:** No tracking issue exists for this fallback. Searched open and closed issues/PRs across `codex`, `client_version`, `manifest`, and `/v1/models`; this is a self-contained compatibility fix and does not map to a reported bug.

### Core Changes
- `src/app/v1/_lib/models/available-models.ts` (+29/-1) — in `handleAvailableModels`, after `authenticateRequest`, branch only for exact `/v1/models` requests in OpenAI format, without an explicit format override, and with a non-empty `client_version`: set a stable weak `ETag` (`W/"cch-codex-bundled-v1"`) and `Cache-Control: no-cache`, return `304` for matching, multi-value, weak-equivalent, or wildcard `If-None-Match`, otherwise return `{"models":[]}`. Ordinary `/v1/models`, Gemini `/v1beta/models`, `x-goog-api-key`, and explicit format-override requests fall through unchanged to the existing aggregator.

### Supporting Changes
- `tests/unit/models/codex-models-manifest-fallback.test.ts` (+173, new) — covers the empty manifest, ETag/cache headers, exact/list/wildcard validators, unchanged OpenAI and Gemini discovery (including `x-goog-api-key` and explicit format overrides), whitespace-only parameters, and authentication failures.

## Scope

This is deliberately a compatibility fallback only. It does not proxy provider-specific manifests, synthesize capability fields, change `/v1/responses/models`, or alter ordinary OpenAI model discovery.

## Breaking Changes

None. The new branch is purely additive and runs only for exact `/v1/models` requests in OpenAI format, without an explicit format override, when `client_version` is present and non-empty; every other model-list request is unchanged. No schema, migration, or exported-signature changes.

### Automated Tests
- [x] Unit tests added (`tests/unit/models/codex-models-manifest-fallback.test.ts`, +173)
- [ ] Integration tests — N/A (single handler branch, fully covered by unit tests)

### Manual Testing
1. `GET /v1/models?client_version=0.144.1` with a valid key → expect `200 {"models":[]}` and `ETag: W/"cch-codex-bundled-v1"`.
2. Repeat with a matching, multi-value, or wildcard `If-None-Match` → expect `304` with an empty body.
3. `GET /v1/models` (no `client_version`) → expect the unchanged OpenAI `{object,data}` response and no `ETag`.
4. `GET /v1beta/models?client_version=0.144.1` → expect unchanged Gemini model discovery.
5. `GET /v1/models?client_version=0.144.1` with `x-goog-api-key` → expect unchanged Gemini model discovery.
6. `GET /v1/models?client_version=0.144.1&format=gemini` (or equivalent header/query override) → expect unchanged explicit provider selection.
7. `GET /v1/models?client_version=0.144.1` with no/invalid credentials → expect `401`, no manifest returned.

### Validation
- `bun run test tests/unit/models/codex-models-manifest-fallback.test.ts tests/unit/proxy/available-models.test.ts tests/unit/models/available-models-auth-outcome.test.ts tests/unit/models/available-models-gemini-key.test.ts` (45 passed)
- `bun run typecheck`
- `bun run lint`
- `bun run format:check`
- `bunx @biomejs/biome@2.5.4 check src/app/v1/_lib/models/available-models.ts tests/unit/models/codex-models-manifest-fallback.test.ts`

## Checklist
- [x] Code follows project conventions
- [x] Self-review completed
- [x] Tests pass locally
- [ ] Documentation updated (if needed)

---
*Description enhanced by Claude AI*

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds a Codex-manifest compatibility branch to `handleAvailableModels`: when a request targets exactly `/v1/models` in OpenAI format (no `x-goog-api-key`, no `/v1beta/` path, no explicit format override) and supplies a non-empty `client_version`, the handler short-circuits to return `{"models":[]}` with a stable weak ETag and `Cache-Control: no-cache` instead of routing through the OpenAI model aggregator. Authentication still runs first, so 401 handling is unaffected.

- **Codex branch gating** (`available-models.ts`): five conditions must all be true — exact path, `responseFormat === "openai"`, no `clientFormatOverride`, non-whitespace `client_version` — before the early return fires; all other requests reach the existing aggregator unchanged.
- **RFC 7232 compliance**: `matchesIfNoneMatch` implements correct weak comparison (strips `W/` from both sides), handles comma-separated validators and wildcards, and attaches ETag/Cache-Control to both the 200 and 304 responses.
- **Test coverage** (`codex-models-manifest-fallback.test.ts`): 8 test cases covering the empty manifest, 304 with exact/multi-value/wildcard validators, Gemini and explicit-override fall-throughs, whitespace `client_version` truncation, and both authentication-failure scenarios.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

Safe to merge — the new branch is strictly additive, authentication is enforced before it fires, and all existing model-list paths are unchanged.

The Codex branch is guarded by five independent conditions, so the risk of accidental activation is low. The matchesIfNoneMatch helper correctly follows RFC 7232 weak-comparison rules (wildcard, comma-separated, and weak-prefixed ETags all tested). ETag and Cache-Control headers are set before both the 200 and 304 responses. The test suite exercises every documented edge case including authentication failures, Gemini fall-throughs, whitespace client_version, and explicit format overrides. No schema changes, no migration, no exported-signature changes.

No files require special attention.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/app/v1/_lib/models/available-models.ts | Adds a matchesIfNoneMatch helper and a Codex-manifest early-return branch in handleAvailableModels; authentication still runs first and all existing paths are unchanged. |
| tests/unit/models/codex-models-manifest-fallback.test.ts | New test file covering the empty manifest, ETag/Cache-Control headers, wildcard and multi-value If-None-Match, Gemini and explicit-override fall-throughs, whitespace client_version, and both auth-failure scenarios. |

</details>

<details open><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Client as Codex CLI / OpenAI Client
    participant H as handleAvailableModels
    participant Auth as authenticateRequest
    participant Agg as Model Aggregator

    Client->>H: "GET /v1/models?client_version=0.144.1"
    H->>Auth: authenticateRequest(c)
    Auth-->>H: "{user, key} or throws 401"
    H->>H: "detectResponseFormat -> openai"
    H->>H: "detectClientFormatOverride -> null"
    H->>H: "path===/v1/models and client_version non-empty?"
    alt Codex manifest branch
        H->>H: set ETag + Cache-Control headers
        alt If-None-Match matches
            H-->>Client: 304 Not Modified + ETag + Cache-Control
        else No match
            H-->>Client: 200 models[] + ETag + Cache-Control
        end
    else Normal aggregator path
        H->>Agg: getAvailableModels
        Agg-->>H: models + providerName
        H-->>Client: 200 OpenAI or Gemini or Anthropic shape
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Client as Codex CLI / OpenAI Client
    participant H as handleAvailableModels
    participant Auth as authenticateRequest
    participant Agg as Model Aggregator

    Client->>H: "GET /v1/models?client_version=0.144.1"
    H->>Auth: authenticateRequest(c)
    Auth-->>H: "{user, key} or throws 401"
    H->>H: "detectResponseFormat -> openai"
    H->>H: "detectClientFormatOverride -> null"
    H->>H: "path===/v1/models and client_version non-empty?"
    alt Codex manifest branch
        H->>H: set ETag + Cache-Control headers
        alt If-None-Match matches
            H-->>Client: 304 Not Modified + ETag + Cache-Control
        else No match
            H-->>Client: 200 models[] + ETag + Cache-Control
        end
    else Normal aggregator path
        H->>Agg: getAvailableModels
        Agg-->>H: models + providerName
        H-->>Client: 200 OpenAI or Gemini or Anthropic shape
    end
```

</a>
</details>

<sub>Reviews (5): Last reviewed commit: ["test(models): stabilize formatter output"](https://github.com/ding113/claude-code-hub/commit/f2eeff3232fd587ef16fcd47546eca98b04da53b) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44957114)</sub>

### CODE REVIEW

**Comment by Brisbanehuang** [comment]
CI note for reviewers:

- Unit, API, and integration jobs pass. The focused models tests pass locally (40 tests), as do typecheck, lint, format check, and `git diff --check`.
- The format failures come from a fresh CI install resolving `@biomejs/biome@2.5.4` through the repository's unpinned `^2` dependency and no committed lockfile. That version requests formatting in 23 existing `dev` files; neither file changed by this PR appears in the reported set. I have intentionally not added unrelated formatting churn to this compatibility fix.
- Docker Build Test failed before the Docker build in `docker/metadata-action@v6`, where the GitHub API returned a full `Unicorn! · GitHub` HTML error page. I attempted to rerun that job, but the fork integration does not have Actions write permission on the upstream repository. A maintainer rerun should be sufficient.

The patch remains limited to the authenticated `client_version` branch and its regression tests.

File: src/app/v1/_lib/models/available-models.ts [hunk 1]
```diff
@@ -544,6 +547,14 @@ export async function handleAvailableModels(c: Context): Promise<Response> {
   try {
     const { user, key } = await authenticateRequest(c);
 
+    if (c.req.query("client_version")?.trim()) {
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/models/available-models.ts]: Fixed in f32db983. The fallback now requires the exact `/v1/models` path in addition to a non-empty `client_version`. I also added a `/v1beta/models?client_version=...` regression case proving that Gemini aggregation remains active. The focused suite passes with 43 tests.

File: src/app/v1/_lib/models/available-models.ts [hunk 2]
```diff
@@ -544,6 +547,14 @@ export async function handleAvailableModels(c: Context): Promise<Response> {
...
+      c.header("ETag", CODEX_MODELS_MANIFEST_ETAG);
+      if (c.req.header("if-none-match") === CODEX_MODELS_MANIFEST_ETAG) {
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/models/available-models.ts]: Fixed in 000e1966. `If-None-Match` now supports `*`, comma-separated validators, and weak comparison for GET requests. Regression coverage includes the exact ETag, a multi-value header, and the wildcard case.

File: src/app/v1/_lib/models/available-models.ts [hunk 3]
```diff
@@ -544,6 +547,14 @@ export async function handleAvailableModels(c: Context): Promise<Response> {
...
+        return c.body(null, 304);
+      }
+      return c.json({ models: [] });
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/models/available-models.ts:573]: Fixed in 000e1966. Both the 200 manifest response and matching 304 response now include `Cache-Control: no-cache`, with assertions in the regression test.

File: tests/unit/models/codex-models-manifest-fallback.test.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+  };
+});
+
+import { handleAvailableModels } from "@/app/v1/_lib/models/available-models";
+import { resolveApiKeyAuthOutcome } from "@/repository/key";
+import { findAllProviders } from "@/repository/provider";
+
+const CODEX_MODELS_ETAG = 'W/"cch-codex-bundled-v1"';
+
+function createApp() {
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply tests/unit/models/codex-models-manifest-fallback.test.ts:48]: I am keeping the expected ETag as an independent test literal. This test locks the public HTTP contract; importing the implementation constant would let an accidental value change update both producer and expectation together, weakening the regression check. The constant remains private because it is not an application API.

File: src/app/v1/_lib/models/available-models.ts [hunk 5]
```diff
@@ -544,6 +557,15 @@ export async function handleAvailableModels(c: Context): Promise<Response> {
   try {
     const { user, key } = await authenticateRequest(c);
 
+    if (c.req.path === "/v1/models" && c.req.query("client_version")?.trim()) {
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/models/available-models.ts]: Fixed in af5cb4c8. `detectResponseFormat()` now runs before the fallback, which additionally requires `responseFormat === "openai"`. A new regression case covers `/v1/models?client_version=...` with `x-goog-api-key` and confirms Gemini aggregation remains active. The focused suite passes with 44 tests.

File: src/app/v1/_lib/models/available-models.ts [hunk 6]
```diff
@@ -543,8 +556,21 @@ export const handleOpenAICompatibleModels = createFixedProviderTypesModelsHandle
 export async function handleAvailableModels(c: Context): Promise<Response> {
   try {
     const { user, key } = await authenticateRequest(c);
-
     const responseFormat = detectResponseFormat(c);
+
+    if (
+      c.req.path === "/v1/models" &&
+      responseFormat === "openai" &&
+      c.req.query("client_version")?.trim()
```
[the remark this answers is not in the document]
Comment (Brisbanehuang) [reply src/app/v1/_lib/models/available-models.ts:566]: Fixed in 3d5e42f9. `detectClientFormatOverride()` now runs before the fallback, which additionally requires `clientFormatOverride === null`. Regression coverage includes `format=gemini`, `api_type=gemini`, and `x-cch-api-type: gemini`; all continue through provider aggregation even with `client_version`. The focused suite passes with 47 tests.

**Comment by Brisbanehuang** [comment]
CI follow-up: a later review revision exposed a Biome 2.4/2.5 formatting disagreement in this PR's parameterized test. Commit `f2eeff32` replaces that syntax with formatter-stable code. Both the repository formatter and CI's resolved `@biomejs/biome@2.5.4` now pass when run directly against the two changed files. Any remaining full-repository format failures are limited to existing `dev` files outside this PR.

**Comment by Brisbanehuang** [comment]
Final status for `f2eeff32`:

- Unit Tests: passed, including v1 management API tests.
- API Tests, Integration Tests, Docker Build Test, and Greptile Review: passed.
- Local focused regression suite: 45 passed; typecheck, lint, repository format check, and `git diff --check` passed.
- CI formatter parity: `@biomejs/biome@2.5.4` passes when run directly against both files changed by this PR.
- The two remaining Code Quality jobs run Biome 2.5.4 across the full repository and report 23 pre-existing `dev` files outside this PR. The changed files are not in that error set.
- All review threads are resolved; the latest review-driven fixes cover `/v1beta`, `x-goog-api-key`, explicit format overrides, standard `If-None-Match`, and cache revalidation.
